src/Microsoft.AspNetCore.Diagnostics.HealthChecks/HealthCheckMiddleware.cs

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using Microsoft.Extensions.Options;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.AspNetCore.Diagnostics.HealthChecks
+{
+    public class HealthCheckMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly HealthCheckOptions _healthCheckOptions;
+        private readonly IHealthCheckService _healthCheckService;
+
+        public HealthCheckMiddleware(RequestDelegate next, IOptions<HealthCheckOptions> healthCheckOptions, IHealthCheckService healthCheckService)
+        {
+            _next = next;
+            _healthCheckOptions = healthCheckOptions.Value;
+            _healthCheckService = healthCheckService;
+        }
+
+        /// <summary>
+        /// Process an individual request.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        public async Task InvokeAsync(HttpContext context)
+        {
+            if (context.Request.Path == _healthCheckOptions.Path)
+            {
+                // TODO: Caching
+                // Get results
+                var result = await _healthCheckService.CheckHealthAsync(context.RequestAborted);
+
+                // Map status to response code
+                switch (result.Status)
+                {
+                    case HealthCheckStatus.Failed:
+                        // REVIEW: Maybe we shouldn't distinguish between failed health checks and unhealthy?
+                        context.Response.StatusCode = StatusCodes.Status500InternalServerError;
```

Comment:
I think we should. This will make it easier for us to support.

Reply:
Not only that it makes it easier to troubleshoot. Are my services going offline cause the coder sucks and wrote a bad healthcheck or is the service actually failing

Reply:
Yep, that's why I put it this way initially. I'll remove the comment, sounds like there's consensus :)